tests/security/static/queries/cpp/openenclave/Dereferences.qll

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+/**
+ * Provides classes and predicates for identifying expressions that are dereferences of a pointer.
+ *
+ * This library detects dereferences that occur locally - for example, in a dereferencing
+ * expression, or as an argument to a library call such as memcpy - as well as dereferences that
+ * occur by passing pointer arguments to calls that ultimately dereference the argument.
+ */
+
+import cpp
+import semmle.code.cpp.controlflow.Dereferenced
+
+/**
+ * A pointer expression `e` that is dereferenced directly by `op`.
+ */
+predicate localDereference(Expr op, Expr e) {
+  e.getActualType().getUnspecifiedType() instanceof PointerType and
+  not op = any(SizeofExprOperator sizeof).getAChild*() and
+  (
+    e = op.(PointerDereferenceExpr).getOperand()
+    or
+    e = op.(FieldAccess).getQualifier()
+    or
+    e = op.(Call).getQualifier()
+    or
+    e = op.(ArrayExpr).getArrayBase()
+    or
+    e = op.(ArrayExpr).getArrayOffset()
+    or
+    localDereference(op, any(PointerArithmeticOperation pOp | e = pOp.getAnOperand()))
```

Comment:
Does this handle `ptr +1` and `1 + ptr` ?

Reply:
Yes it is handled by PointerArithmeticOperation expression node.
>> localDereference(op, any(PointerArithmeticOperation pOp | e = pOp.getAnOperand()))